### ISSUE TITLE
ncrcat fails to concat icebergs restarts 

### ISSUE THREAD

**Issue by nikizadehgfdl** [NOAA-GFDL/icebergs#10]
Using dev/master for icebergs project I get restart files that cannot be combined and hence the model cannot restart.

```
% cd /lustre/f1/Niki.Zadeh/work/ulm_201505_mom6_2015.07.20/OM4_SIS2_baseline_1x0m10d_2997pe1.o7143211/output.stager/lustre/f1/Niki.Zadeh/ulm_201505_mom6_2015.07.20_0/OM4_SIS2_baseline/ncrc2.intel-prod/1x0m10d_2997pe1/restart/19000111

% ncrcat icebergs.res.nc.0000 icebergs.res.nc.0001 icebergs.res.nc.0002 icebergs.res.nc.0003 icebergs.res.nc.0004 icebergs.res.nc.0005 icebergs.res.nc.0006 icebergs.res.nc.0007 icebergs.res.nc.0008 icebergs.res.nc.0009 icebergs.res.nc.0010 icebergs.res.nc.0011 icebergs.res.nc.0012 icebergs.res.nc.0013 icebergs.res.nc.0014 icebergs.res.nc.0015 icebergs.res.nc.0016 icebergs.res.nc.0017 icebergs.res.nc.0018 icebergs.res.nc.0019 icebergs.res.nc.0020 icebergs.res.nc.0021 icebergs.res.nc.0022 icebergs.res.nc.0023 icebergs.res.nc.0024 icebergs.res.nc.0025 icebergs.res.nc.0026 icebergs.res.nc.0028 icebergs.res.nc.0029 icebergs.res.nc.0030 icebergs.res.nc.0031 icebergs.res.nc.0032 icebergs.res.nc.0033 icebergs.res.nc.0034 icebergs.res.nc.0035 icebergs.res.nc.0036 icebergs.res.nc.0037 icebergs.res.nc.0038 icebergs.res.nc.0039 icebergs.res.nc.0040 icebergs.res.nc.0041 icebergs.res.nc.0042 icebergs.res.nc.0043 icebergs.res.nc.0044 icebergs.res.nc
ncrcat: ERROR nco_lmt_sct_mk() reports record variable exists and is size zero, i.e., has no records yet.
ncrcat: HINT: Perform record-oriented operations only after file has valid records.
ncrcat: cnt < 0 in nco_lmt_sct_mk()
```

**Comment by adcroft** [comment]
I looked at those files and every single one is empty which means there
were no bergs generated/surviving. Evidently we need to stop mpp from
writing files where there are no bergs.

-A.

## 

Dr Alistair Adcroft    (Alistair.Adcroft@noaa.gov)
Princeton University           Tel: (609) 987-5073
NOAA/GFDL, 201 Forrestal Road, Princeton, NJ 08540

**Comment by jwdGFDL** [comment]
This will entail a rather sticky change to the read operation as there's 
currently a sanity check that the complete set of files is there (i.e. 
the number of files is consistent with that expected by the layout).

Suggest that the simpler approach is to modify the combine processing to 
ignore files with 0 elements via a simple check on the metadata. In 
fact, seems to me that ncrcat should do that as a native operation.

**Comment by adcroft** [comment]
I agree about ncrcat but in this instance it might be complaining not about
individual files but that there are no records at all. We should find out
if ncrcat complains if there are some bergs in at least one file.

-A.

## 

Dr Alistair Adcroft    (Alistair.Adcroft@noaa.gov)
Princeton University           Tel: (609) 987-5073
NOAA/GFDL, 201 Forrestal Road, Princeton, NJ 08540

**Comment by nikizadehgfdl** [comment]
ncrcat complains if there are any files with no record among the input files. E.g., 0001 and 0002 below have no bergs:

```
gaea4: /lustre/f1/Niki.Zadeh/work/ulm_201505_mom6_2015.07.20/MOM6_GOLD_SIS2_bergs_1x0m20d_64pe.o2491150/RESTART % ncrcat icebergs.res.nc.0000 icebergs.res.nc.0001 icebergs.res.nc.0002 icebergs.res.nc.0003 icebergs.res.nc
ncrcat: ERROR Size of dimension "i" is 0 in input file, but must be > 0 in order to apply limits.
gaea4: /lustre/f1/Niki.Zadeh/work/ulm_201505_mom6_2015.07.20/MOM6_GOLD_SIS2_bergs_1x0m20d_64pe.o2491150/RESTART % ncrcat icebergs.res.nc.0000 icebergs.res.nc.0003 icebergs.res.nc 
OK
```

**Comment by nikizadehgfdl** [comment]
duplicate of #2

**Comment by sternalon** [comment]
Hi All
It's strange that we are suddenly encountering this issue, since we have been using iceberg restarts for a while already, and combining restarts with no icebergs...
Did we change something recently that make it stop working?

I am away today, but I can take a look at this when I'm back on Monday. 

Best,
Alon

**Comment by nikizadehgfdl** [comment]
Good question Alon. You have not seen this issue because you are probably using a ice_io_layout=1,0  which would produce only one iceberg restart file and the issue does not arise. This issue has nothing to do with your recent developments, so have a peaceful weekend :)

But this is real issue for the workflow. A little background to make it clear:
1. In the xml workflow the icebergs restarts are ALWAYS combined via ncrcat and the  uncombined files are removed at the end of a segment and BEFORE the next segment starts. That means we have never tested restarting from uncombined icebergs restarts before!
2. Before new icebergs I/O developments only PEs which had icebergs in them would write a restart, so ncrcat was happy since there were no "empty" restart files in its input list.
3. The reason we did (1) above was that FRE does not know how to combine icebergs restarts and tries to treat them with the generic mppnccombine which cause the output stager of "regression" runs to fail (not the "production" runs since the restarts are not combined then anyway). 

I think we need to fix FRE to treat iceberg restarts differently, just as it does the land restarts, and get rid of the step 1 above in the xml.  Putting the logic to find "empty" restarts in the xml is too much and counter productive (since we want them not combined at the end of every segment).

**Comment by nikizadehgfdl** [comment]
I put an ugly workaround in the xml postProcess section to bypass this issue until it is fixed properly in the codebase.

```
         #Find non-empty iceberg restarts and combine them AND then remove uncombined restarts
         #This should be handled properly by FRE
         #ncrcat icebergs.res.nc.* icebergs.res.nc
         set non_empty_restarts = ""
         foreach file ( `ls icebergs.res.nc.????` )
            if ( `ncdump -h $file | grep UNLIMITED | awk '{gsub(/\(/," ");print $6}'` ) set non_empty_restarts = "$non_empty_restarts $file"
         end
         ncrcat $non_empty_restarts icebergs.res.nc
```

**Comment by jwdGFDL** [comment]
Niki,
     While I think it's fairly easy to prevent a restart from being 
written when there's no data in the global domain, In the case where 
there are icebergs somewhere but a particular I/O domain is empty, the 
current infrastructure writes a file anyway because having a complete 
file set is one of the expectations for the read. I suppose we could 
introduce optional flags to relax this behavior. But a completely 
consistent set of logic is probably really non-trivial, especially if 
we're going to want to checksum the fields (and given Larry's recent 
experience, I expect that we do).

The net is that filtering restarts with no data doesn't seem all that 
horrendous to me.

On 08/03/2015 01:33 PM, Niki Zadeh wrote:

> I put an ugly workaround in the xml postProcess section to bypass this 
> issue until it is fixed properly in the codebase.
> 
> |#Find non-empty iceberg restarts and combine them AND then remove 
> uncombined restarts #This should be handled properly by FRE #ncrcat 
> icebergs.res.nc.\* icebergs.res.nc set non_empty_restarts = "" foreach 
> file ( `ls icebergs.res.nc.????` ) if ( `ncdump -h $file | grep 
> UNLIMITED | awk '{gsub(/\(/," ");print $6}'` ) set non_empty_restarts 
> = "$non_empty_restarts $file" end ncrcat $non_empty_restarts 
> icebergs.res.nc |
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/NOAA-GFDL/icebergs/issues/10#issuecomment-127339713.

## 

Jeff Durachta
Engineering Lead for Modeling Services
NOAA Geophysical Fluid Dynamics Lab
Forrestal Campus, Princeton University
201 Forrestal Road
Princeton, NJ  08540
Office: +1-609-987-5054

**Comment by nikizadehgfdl** [comment]
Added the following logic to the xml to avoid ncrcating empty restarts.

```
         set non_empty_restarts = ""
         foreach file ( `ls icebergs.res.nc.????` )
            if ( `ncdump -h $file | grep UNLIMITED | awk '{gsub(/\(/," ");print $6}'` ) set non_empty_restarts = "$non_empty_restarts $file"
         end
         ncrcat $non_empty_restarts icebergs.res.nc

```
